packages/react-dom/src/server/ReactPartialRenderer.js

```diff
@@ -585,6 +590,27 @@ class ReactDOMServerRenderer {
       if (nextChild === null || nextChild === false) {
         return '';
       } else if (!React.isValidElement(nextChild)) {
+        if (nextChild != null && nextChild.$$typeof != null) {
+          // Catch unexpected special types early.
+          const $$typeof = nextChild.$$typeof;
+          invariant(
```

Comment:
I used a switch here at first but Flow and ESLint started fighting over presence of `break`s (which are technically unreachable). This is shorter anyway. It's not a hot path (always errs).